Update letterSpacing and wordSpacing to accept String

yiyix opened the pull request:
<!--
Thank you for contributing to the HTML Standard! Please describe the change you are making and complete the checklist below if your change is not editorial.
-->

- [x] At least two implementers are interested (and none opposed):
   - [ ] Chrome
   - [ ] Firefox
   
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   - [x] https://github.com/web-platform-tests/wpt/pull/31735
   - [x] https://github.com/web-platform-tests/wpt/pull/32302
   
- Implementation bugs are filed:
  - [x] Chrome: https://bugs.chromium.org/p/chromium/issues/detail?id=1263632 
  - [x] Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1708327
  - [x] Safari: https://bugs.webkit.org/show_bug.cgi?id=225181

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/7378/canvas.html" title="Last updated on Dec 14, 2021, 6:09 AM UTC (feb1a68)">/canvas.html</a>  ( <a href="https://whatpr.org/html/7378/3ad5159...feb1a68/canvas.html" title="Last updated on Dec 14, 2021, 6:09 AM UTC (feb1a68)">diff</a> )


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/7378/canvas.html" title="Last updated on Jan 28, 2022, 7:59 PM UTC (5684404)">/canvas.html</a>  ( <a href="https://whatpr.org/html/7378/e8e499e...5684404/canvas.html" title="Last updated on Jan 28, 2022, 7:59 PM UTC (5684404)">diff</a> )

annevk commented:
Thanks for creating this PR @yiyix.

As far as I know @Kaiido does not speak for Mozilla or Firefox. @jdashg or I do (at times).

Also, this change isn't sufficient. You need to invoke the CSS parser as we do for other canvas members. And we also need to serialize the value in some manner for the getters.

annevk commented:
cc @whatwg/canvas

Kaiido reviewed (changes requested):
Yes, I always speak (maybe too much) only for myself as a web-author, I don't have any affiliation with any vendor.

Regarding the changes I still find the discrepancy between `font` being converted to `px` and these values kept as relative `<length>` a bit surprising (`"2vw"` may represent two different values entirely just by resizing the window), but overall I don't have strong opinions.

yiyix commented:
> Thanks for creating this PR @yiyix.
> 
> As far as I know @Kaiido does not speak for Mozilla or Firefox. @jdashg or I do (at times).
Thank you for the clarification. What do you think about this change, Define letterSpacing and wordSpacing in terms of String? Or do you think it's better to keep as it is or accept doubles and strings? 

> Also, this change isn't sufficient. You need to invoke the CSS parser as we do for other canvas members. And we also need to serialize the value in some manner for the getters.
I will re-write this part. Thank you

Kaiido commented on source:
```diff
@@ -62041,8 +62041,9 @@ transform. ack Shaun Morris. -->
 
-    <p>Can be set, to change spacing between characters. Postive values spreads characters further
-    apart, while negative values brings them closer together. The default is 0.</p>
+    <p>Can be set, to change spacing between characters. The syntax is same as CSS
+    <span>&lt;length></span>; values that cannot be parsed as CSS <span>'letter-spacing'</span> property are ignored. The
+    default is "0px".</p>
```
```suggestion
    <p>Can be set, to change the spacing between characters. The syntax is the same as for the CSS
    <span>&lt;length></span> dimension; values that cannot be parsed as a CSS
    <span>'letter-spacing'</span> value are ignored. The default is "0px".</p>
```

I am not a native English speaker, so take the "the spacing" suggestion with a grain of salt.
Also I'm not sure "dimension" is the correct term here; [the &lt;def>](https://html.spec.whatwg.org/multipage/infrastructure.html#length-2) doesn't have a proper one, but there should probably be something here.

Finally see below comments where I proposed to change from "parsing as 'letter-spacing'" to "parsing as '&lt;length>'", not sure if this note should also reflect that or if it' fine like that.

yiyix replied:
dimension.... I definitely used the wrong work, will update. 

The reason I put parse as '<length>' is because css `letter-spacing` also takes value such as normal, initial, etc ([link](https://developer.mozilla.org/en-US/docs/Web/CSS/letter-spacing)) I will wait for some suggestions from @domenic domenic.

domenic replied:
I think "values that cannot be parsed as a CSS '\<length>' value are ignored" would make the most sense here.

Kaiido commented on source:
```diff
@@ -62090,8 +62091,9 @@ transform. ack Shaun Morris. -->
 
-    <p>Can be set, to change spacing between words. Postive values spreads words further apart,
-    while negative values brings them closer together. The default is 0.</p>
+    <p>Can be set, to change spacing between characters. The syntax is same as CSS
+    <span>&lt;length></span>; values that cannot be parsed as CSS <span>'word-spacing'</span> property are ignored. The
+    default is "0px".</p>
```
```suggestion
    <p>Can be set, to change the spacing between words. The syntax is the same as for the CSS
    <span>&lt;length></span> dimension; values that cannot be parsed as a CSS
    <span>'word-spacing'</span> value are ignored. The default is "0px".</p>
```
(With the same grain-of-salt as for the one above)

Kaiido commented on source:
```diff
@@ -62245,12 +62247,13 @@ worker.postMessage(offscreenCanvas, [offscreenCanvas]);</code></pre>
-  <code>CanvasTextDrawingStyles</code> interface is created,
-  the <code data-x="dom-context-2d-letterSpacing">letterSpacing</code> attribute must
-  initially have the value 0.</p>
+  data-x="dom-context-2d-letterSpacing">letterSpacing</code></dfn> IDL attribute, on getting, must return
+  the <span data-x="serializing a CSS value">serialized form</span> of the current value. On setting, must
+  be a <span>&lt;length></span> value and <span
+  data-x="parse something according to a CSS grammar">parsed as a CSS &lt;'letter-spacing'> property</span>. If the new value is syntactically incorrect, then it must
+  be ignored, without assigning a new <code data-x="dom-context-2d-letterSpacing">letterSpacing</code>
+  value. When the object implementing the <code>CanvasTextDrawingStyles</code> interface is created,
+  the <code data-x="dom-context-2d-letterSpacing">letterSpacing</code> attribute must initially have the value "0px".</p>
```
```suggestion
  data-x="dom-context-2d-letterSpacing">letterSpacing</code></dfn> IDL attribute, on getting, must
  return the <span data-x="serializing a CSS value">serialized form</span> of the current value.
  On setting, the value must be <span data-x="parse something according to a CSS grammar">parsed as
  a CSS &lt;length> value</span>. If the return value is failure then it must be ignored, without
  assigning a new <code data-x="dom-context-2d-letterSpacing">letterSpacing</code> value. When the
  object implementing the <code>CanvasTextDrawingStyles</code> interface is created, the <code
  data-x="dom-context-2d-letterSpacing">letterSpacing</code> attribute must initially have the
  value "0px".</p>
```

Here I am not sure, is it ok to only parse as `<length>`? I guess the original intent was to avoid values like `"normal"`or `"inherit", right?
For `font` the current text says 

> must be parsed as a CSS <'font'> value (but without supporting property-independent style sheet syntax like 'inherit')

but here, I have the hope that simply parsing as `<number>` would do.

Kaiido commented on source:
```diff
@@ -62281,12 +62284,13 @@ worker.postMessage(offscreenCanvas, [offscreenCanvas]);</code></pre>
-  <code>CanvasTextDrawingStyles</code> interface is created,
-  the <code data-x="dom-context-2d-wordSpacing">wordSpacing</code> attribute must
-  initially have the value 0.</p>
+  data-x="dom-context-2d-wordSpacing">wordSpacing</code></dfn> IDL attribute, on getting, must return
+  the <span data-x="serializing a CSS value">serialized form</span> of the current value. On setting, must
+  be a <span>&lt;length></span> value and <span
+  data-x="parse something according to a CSS grammar">parsed as a CSS &lt;'word-spacing'> property</span>. If the new value is syntactically incorrect, then it must
+  be ignored, without assigning a new <code data-x="dom-context-2d-letterSpacing">wordSpacing</code>
+  value. When the object implementing the <code>CanvasTextDrawingStyles</code> interface is created,
+  the <code data-x="dom-context-2d-wordSpacing">wordSpacing</code> attribute must initially have the value "0px".</p>
```
```suggestion
  data-x="dom-context-2d-wordSpacing">wordSpacing</code></dfn> IDL attribute, on getting, must
  return the <span data-x="serializing a CSS value">serialized form</span> of the current value.
  On setting, the value must be <span data-x="parse something according to a CSS grammar">parsed as
  a CSS &lt;length> value</span>. If the return value is failure then it must be ignored, without
  assigning a new <code data-x="dom-context-2d-wordSpacing">wordSpacing</code> value. When the
  object implementing the <code>CanvasTextDrawingStyles</code> interface is created, the <code
  data-x="dom-context-2d-wordSpacing">wordSpacing</code> attribute must initially have the
  value "0px".</p>
```
(same note as above)

Kaiido reviewed (changes requested):
Sorry for this quite sketchy review, I am not sure about all my proposed changes and a review of the review is required.

domenic commented on source:
```diff
@@ -62245,12 +62247,13 @@ worker.postMessage(offscreenCanvas, [offscreenCanvas]);</code></pre>
+  data-x="parse something according to a CSS grammar">parsed as a CSS &lt;'letter-spacing'> property</span>. If the new value is syntactically incorrect, then it must
+  be ignored, without assigning a new <code data-x="dom-context-2d-letterSpacing">letterSpacing</code>
+  value. When the object implementing the <code>CanvasTextDrawingStyles</code> interface is created,
+  the <code data-x="dom-context-2d-letterSpacing">letterSpacing</code> attribute must initially have the value "0px".</p>
```
I think we need to do some extra legwork here to make things more rigorous, now that they are more complicated.

So:

* Say something like `Objects that implement the CanvasTextDrawingStyles interface have an associated letter spacing, which is a CSS <length> value. It is initially the result of parsing "0px"`.
* Define "the letterSpacing getter steps" which return the serialization of this's letter spacing.
* Define "the letterSpacing setter steps" which, if I understand the intent correctly, is something like this:
  
  1. Let _parsed_ be the result of parsing the given value as a CSS \<length> value.
  2. If _parsed_ is failure, then return.
  3. Set this's letter spacing to _parsed_.

It might be worth doing this first for just letter spacing, and then once we get that perfect we can also do word spacing, which should be very similar. Or you can do them both at once; it's up to you.

yiyix replied:
Do we really need "the letterSpacing getter steps"? It basically serialize the current CSS <length> value. It's seems to be easy to explain. I updated this part without adding a section for "the letterSpacing getter steps".

domenic replied:
We do, but it's simpler than what you have: see the template here. Sorry, the template you should follow is more like the following: https://html.spec.whatwg.org/#dom-context-2d-fillstyle The getter/setter steps are supposed to replace the above definition.

domenic reviewed (changes requested):
Overall it looks like you have the right idea to me, and have found the most relevant algorithms for parsing and serializing.

Note that this parse/serialize architecture means that

```js
ctx.letterSpacing = "5PX";
console.assert(ctx.letterSpacing === "5px"); // lowercased!!
```

will hold. That seems consistent with what we've specified elsewhere, but is worth writing web platform tests for.

yiyix commented:
> 



> Overall it looks like you have the right idea to me, and have found the most relevant algorithms for parsing and serializing. We just need to tighten up the language.
> 
> Note that this parse/serialize architecture means that
> 
> ```js
> ctx.letterSpacing = "5PX";
> console.assert(ctx.letterSpacing === "5px"); // lowercased!!
> ```
> 
> will hold. That seems consistent with what we've specified elsewhere, but is worth writing web platform tests for.

Good point! The change is now under review.

domenic commented on source:
```diff
@@ -62245,12 +62245,23 @@ worker.postMessage(offscreenCanvas, [offscreenCanvas]);</code></pre>
+   <li><p>Let <var>parsed</var> be the result of <span data-x="parse something according to a CSS
+   grammar">parsing the given value as a CSS &lt;length></span>.</p></li>
+
+   <li><p>If <var>parsed</var> is <var>failure</var>, then return.</p></li>
```
There is no variable declared as `<var>failure</var>`; this should just be `failure`

domenic commented on source:
```diff
@@ -62245,12 +62245,23 @@ worker.postMessage(offscreenCanvas, [offscreenCanvas]);</code></pre>
+
+   <li><p>If <var>parsed</var> is <var>failure</var>, then return.</p></li>
+
+   <li><p>Set <span>this</span>'s <code data-x="dom-context-2d-letterSpacing">letterSpacing</code>
```
You cannot set `dom-context-2d-letterSpacing` here. You need to declare a separate concept, similar to https://html.spec.whatwg.org/#concept-canvasfillstrokestyles-fill-style

domenic commented on source:
```diff
@@ -62244,13 +62244,51 @@ worker.postMessage(offscreenCanvas, [offscreenCanvas]);</code></pre>
-  initially have the value 0.</p>
+  <p>Objects implement the <code>CanvasTextDrawingStyles</code> interface have attributes and methods (defined in this section) that control the spacing between letters and words.</p>
+
+  <p>Such objects have associated <dfn data-x="concept-CanvasTextDrawingStyles-letter-spacing">letter spacing</dfn> and <dfn data-x="concept-CanvasTextDrawingStyles-word-spacing">word spacing</dfn> values, which are CSS <span>&lt;length></span> values. Initially, both must be the result of <span data-x="serializing a CSS value">serialized form</span> of "0px".</p>
```
I think it should be the result of parsing "0px" as a `<length>`, not serializing.

yiyix replied:
updated.

domenic commented on source:
```diff
@@ -62244,13 +62244,51 @@ worker.postMessage(offscreenCanvas, [offscreenCanvas]);</code></pre>
+
+  <p>Such objects have associated <dfn data-x="concept-CanvasTextDrawingStyles-letter-spacing">letter spacing</dfn> and <dfn data-x="concept-CanvasTextDrawingStyles-word-spacing">word spacing</dfn> values, which are CSS <span>&lt;length></span> values. Initially, both must be the result of <span data-x="serializing a CSS value">serialized form</span> of "0px".</p>
+
+  <p>The <dfn data-x="dom-context-2d-letterSpacing">letterSpacing</dfn> getter steps are:</p>
```
Add `<code>` after the `<dfn>` and `</code>` before the `</dfn>`, here and below.

yiyix replied:
Updated.

domenic commented on source:
```diff
@@ -62244,13 +62244,51 @@ worker.postMessage(offscreenCanvas, [offscreenCanvas]);</code></pre>
+  <p>Such objects have associated <dfn data-x="concept-CanvasTextDrawingStyles-letter-spacing">letter spacing</dfn> and <dfn data-x="concept-CanvasTextDrawingStyles-word-spacing">word spacing</dfn> values, which are CSS <span>&lt;length></span> values. Initially, both must be the result of <span data-x="serializing a CSS value">serialized form</span> of "0px".</p>
+
+  <p>The <dfn data-x="dom-context-2d-letterSpacing">letterSpacing</dfn> getter steps are:</p>
+  <ol>
```
You can collapse this into a single sentence:

> The letterSpacing getter steps are to return the serialized form of this's letter spacing.

yiyix replied:
updated for word spacing as well.

domenic commented on source:
```diff
@@ -62244,13 +62244,51 @@ worker.postMessage(offscreenCanvas, [offscreenCanvas]);</code></pre>
+   <li><p>Set <span>this</span>'s <span data-x="concept-CanvasTextDrawingStyles-letter-spacing">letter spacing</span>
+   to <var>parsed</var>.</p></li>
+
+   <li><p>Return.</p></li>
```
No need for extra return at the end, here and below.

yiyix replied:
Updated.

domenic commented on source:
```diff
@@ -62244,13 +62244,51 @@ worker.postMessage(offscreenCanvas, [offscreenCanvas]);</code></pre>
-  <code>CanvasTextDrawingStyles</code> interface is created,
-  the <code data-x="dom-context-2d-letterSpacing">letterSpacing</code> attribute must
-  initially have the value 0.</p>
+  <p>Objects implement the <code>CanvasTextDrawingStyles</code> interface have attributes and methods (defined in this section) that control the spacing between letters and words.</p>
```
Just attributes, I think, right? Also you can probably merge this with the next paragraph since they seem to be introducing the same concept.

yiyix replied:
By method, I mean those getters and setters. Now I read it, maybe just attributes are enough.

domenic reviewed (commented):
This looks great. A few style issues and one more substantial issue about the initial value, but after that this is about ready to merge. (I can do a final formatting pass.)

domenic reviewed (commented):
OK, this all looks good! But I realized we forgot to update a couple other parts of the spec which currently refer to `letterSpacing`/`wordSpacing` to instead refer to "letter spacing"/"word spacing" (i.e., to refer to the underlying `<span data-x="concept-CanvasTextDrawingStyles-letter-spacing">letter spacing</span>` private field).

I think the following two places need updating:

- https://html.spec.whatwg.org/#the-canvas-state:dom-context-2d-letterspacing
- https://html.spec.whatwg.org/#text-styles:dom-context-2d-letterspacing-3

After that I think we should be good!

yiyix commented:
> OK, this all looks good! But I realized we forgot to update a couple other parts of the spec which currently refer to `letterSpacing`/`wordSpacing` to instead refer to "letter spacing"/"word spacing" (i.e., to refer to the underlying `<span data-x="concept-CanvasTextDrawingStyles-letter-spacing">letter spacing</span>` private field).
> 
> I think the following two places need updating:
> 
> * https://html.spec.whatwg.org/#the-canvas-state:dom-context-2d-letterspacing
Updated
> * https://html.spec.whatwg.org/#text-styles:dom-context-2d-letterspacing-3
Updated
> 
> After that I think we should be good!

This version definitely looks much better than the previous ones. Thank you so much! This should conclude our spec changes for Canvas new API

Kaiido commented:
> I think the following two places need updating:
> 
>    - https://html.spec.whatwg.org/#the-canvas-state:dom-context-2d-letterspacing

Is it so? This section gets the "attributes" values, linking to the IDL attributes' getters sounds correct, no? Otherwise I guess this should be moved in a later item like it's done for "The current dash list".  
And if this is indeed the correct action, then should we also update `strokeStyle`, `fillStyle`, `globalCompositeOperation`, and `filter` *(maybe in an other PR)*?

domenic commented:
Yes @Kaiido, this is part of a general program to update the spec to stop relying on JavaScript-modifiable getters and instead get their backing values.
